x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,341 @@
+        }
+
+        if (numConcurrentReads == 0) {
+            // We sneak peek if there is any thing new in the leader.
+            // If there is we will happily accept
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] peek read [{}]", params.getFollowShardId(), numConcurrentReads, from);
+            sendShardChangesRequest(from, maxReadSize, from);
+        }
+    }
+
+    private boolean hasReadBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentReads >= params.getMaxConcurrentReadBatches()) {
+            LOGGER.trace("{} no new reads, maximum number of concurrent reads have been reached [{}]",
+                params.getFollowShardId(), numConcurrentReads);
+            return false;
+        }
+        if (buffer.size() > params.getMaxWriteBufferSize()) {
+            LOGGER.trace("{} no new reads, buffer limit has been reached [{}]", params.getFollowShardId(), buffer.size());
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized void coordinateWrites() {
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            long sumEstimatedSize = 0L;
+            Translog.Operation[] ops = new Translog.Operation[Math.min(params.getMaxBatchOperationCount(), buffer.size())];
+            for (int i = 0; i < ops.length; i++) {
+                ops[i] = buffer.remove();
+                sumEstimatedSize += ops[i].estimateSize();
+                if (sumEstimatedSize > params.getMaxBatchSizeInBytes()) {
+                    ops = Arrays.copyOf(ops, i + 1);
+                    break;
+                }
+            }
+            numConcurrentWrites++;
+            LOGGER.trace("{}[{}] write [{}/{}] [{}]", params.getFollowShardId(), numConcurrentWrites, ops[0].seqNo(),
+                ops[ops.length - 1].seqNo(), ops.length);
+            sendBulkShardOperationsRequest(ops);
+        }
+    }
+
+    private boolean hasWriteBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentWrites >= params.getMaxConcurrentWriteBatches()) {
+            LOGGER.trace("{} maximum number of concurrent writes have been reached [{}]",
+                params.getFollowShardId(), numConcurrentWrites);
+            return false;
+        }
+        return true;
+    }
+
+
+    private void sendShardChangesRequest(long from, int maxOperationCount, long maxRequiredSeqNo) {
+        innerSendShardChangesRequest(from, maxOperationCount,
+            response -> {
+                retryCounter.set(0);
+                handleReadResponse(from, maxOperationCount, maxRequiredSeqNo, response);
+            },
+            e -> handleFailure(e, () -> sendShardChangesRequest(from, maxOperationCount, maxRequiredSeqNo)));
+    }
+
+    private void handleReadResponse(long from, int maxOperationCount, long maxRequiredSeqNo, ShardChangesAction.Response response) {
```

Comment:
What do you think of:
```
    private void handleReadResponse(long from, int maxOperationCount, long maxRequiredSeqNo, ShardChangesAction.Response response) {
        maybeUpdateMapping(response.getIndexMetadataVersion(), () -> {
            synchronized (ShardFollowNodeTask.this) {
                globalCheckpoint = Math.max(globalCheckpoint, response.getGlobalCheckpoint());
                final long newMinRequiredSeqNo;
                if (response.getOperations().length == 0) {
                    newMinRequiredSeqNo = from;
                } else {
                    assert response.getOperations()[0].seqNo() == from : 
                        "first operation is not what we asked for. From is [" + from + "], got " + response.getOperations()[0];
                    buffer.addAll(Arrays.asList(response.getOperations()));
                    final long maxSeqNo = response.getOperations()[response.getOperations().length - 1].seqNo();
                    assert maxSeqNo==
                        Arrays.stream(response.getOperations()).mapToLong(Translog.Operation::seqNo).max().getAsLong();
                    newMinRequiredSeqNo = maxSeqNo + 1;
                    // update last requested seq no as we may have gotten more than we asked for and we don't want to ask it again.
                    lastRequestedSeqno = Math.max(lastRequestedSeqno, maxSeqNo);
                    assert lastRequestedSeqno <= globalCheckpoint: 
                        "lastRequestedSeqno [" + lastRequestedSeqno + "] is larger than the global checkpoint [" + globalCheckpoint + "]";
                    coordinateWrites();
                }
                
                if (newMinRequiredSeqNo < maxRequiredSeqNo) {
                    int newSize = (int) (maxRequiredSeqNo - newMinRequiredSeqNo) + 1;
                    LOGGER.trace("{} received [{}] ops, still missing [{}/{}], continuing to read...",
                        params.getFollowShardId(), response.getOperations().length, newMinRequiredSeqNo, maxRequiredSeqNo);
                    sendShardChangesRequest(newMinRequiredSeqNo, newSize, maxRequiredSeqNo);
                } else {
                    // read is completed, decrement
                    numConcurrentReads--;
                    if (response.getOperations().length == 0 && globalCheckpoint == lastRequestedSeqno)  {
                        // we got nothing and we have no reason to believe asking again well get us more, treat shard as idle and delay
                        // future requests
                        LOGGER.trace("{} received no ops and no known ops to fetch, scheduling to coordinate reads", 
                            params.getFollowShardId());
                        scheduler.accept(idleShardChangesRequestDelay, this::coordinateReads);
                    } else {
                        coordinateReads();
                    }
                }
            }
        });
    }
```

PS - note the difference in handling of `lastRequestedSeqno` - I think the way you had it had a bug.